### ISSUE TITLE
Make the same highlighting for structure methods and types with identical names

### ISSUE THREAD

**Issue by Vacym** [worlpaker/go-syntax#1]
## Problem

When naming a structure method with the same name as a type name, the method is highlighted as a type name when it is called. While it is not necessarily bad practice to give a method the same name as a type, this can lead to confusion and reduced code readability.

## Illustrating problem

```go
type Test struct{}

func (t Test) Testing() error {
	t.normalName()
	t.error()
	t.int()
	t.any()

	return nil
}

func (t Test) normalName() {}

func (t Test) error() {}

func (t Test) int() {}

func (t Test) any() {}
```

![image](https://user-images.githubusercontent.com/65008219/229570007-887107e1-38f2-485c-9864-ad934c0f1aa2.png)

Using the **Ayu Mirage** theme, all function calls should be the same color as `t.normalName()`.

## Motivation

The proposed change aims to improve code readability and clarity by ensuring that method calls with the same name as a type are highlighted as normal function calls, rather than as type names.

## Reason

In `the go.tmLanguage.json` file, a regular expression is used to identify the type names in the code. For example, this is the pattern for the `error` type:
```json
"storage_types": {
	"patterns": [
		{
			"match": "\\berror\\b",
			"name": "storage.type.error.go"
		}
	]
}
```

However, this pattern fails to distinguish between method calls and type names when they share the same name. As a result, method calls with the same name as a type are highlighted as type names.

## Possible solution

One possible solution is to modify the regular expression in the `storage_types` section of `go.tmLanguage.json` to check for a dot (`.`) preceding the type name. This would ensure that the regular expression only matches type names that are not part of a method call. Here's an example of how this could be done for the `error` type:

```json
"storage_types": {
	"patterns": [
		{
			"match": "(?<!\\.)\\berror\\b",
			"name": "storage.type.error.go"
		}
	]
}
```

By adding `(?<!\\.)` before the type name, the regular expression ensures that the type name is not preceded by a dot, which indicates that it is part of a method call.

This solution can be applied to all type names in `go.tmLanguage.json` to ensure that method calls with the same name as a type are highlighted as normal function calls.

**Comment by worlpaker** [comment]
Great feedback! Thank you for your time and contribution. 
I agree with you, all support functions should be the same color.

While your solution is great, I believe it is unnecessary to include type-declarations in the [Support Functions](https://github.com/worlpaker/go-syntax/blob/c4557358ec1ad78650592d65eb886d0a57d07bd6/syntaxes/go.tmLanguage.json#L1336).
Probably, I forgot to remove it in production.

After removing it:
**Ayu Mirage Theme**

![fix_#1](https://user-images.githubusercontent.com/34960716/229621278-7772f41a-ee3e-4571-bc47-1d84ff3cc707.png)

This has been fixed in v0.2.5 thanks to you, please take a look.

**Comment by Vacym** [comment]
This solution does work. But removing `type-declarations` also removes _type_ highlighting at type conversion.

**Ayu Mirage Theme**

---

v0.2.4

![image](https://user-images.githubusercontent.com/65008219/229634562-d230c279-6e7b-4080-9542-88b752887e15.png)
---

v0.2.5

![image](https://user-images.githubusercontent.com/65008219/229635009-5c52b013-0830-4b91-b9ed-bfe34a8695e1.png)
---

If this behavior is expected, there is no problem.
Although, to be honest, I'm not sure if it's necessary to highlight type conversion like _type_

**Comment by worlpaker** [comment]
Another great feedback. I missed that bug in the last version, thanks for the follow-up.
I believe this extension also should help developers to read code easily.
So it will be better to highlight code according to use.

**v0.2.6 - Ayu Mirage Theme**
![fixed_#1](https://user-images.githubusercontent.com/34960716/229647646-9c5f1b2a-a6e6-43bd-9ee9-caa5be49a2a2.png)

**Comment by Vacym** [comment]
Oh, I'm sorry. Maybe I misspoke. I meant that _type_ highlighting is missing at any type conversion, not just at slice byte conversion. It's just that this is the only type conversion example I found in `semantic_tokens.go`, so I attached it.

Here's another type conversion example with **Ayu Mirage Theme**

v0.2.4

![image](https://user-images.githubusercontent.com/65008219/229681124-57856292-3ac9-452c-9b18-a406fe93c223.png)
---
v0.2.6

![image](https://user-images.githubusercontent.com/65008219/229681209-6e9c9ab3-e0a2-40d1-a960-66ca17c041c3.png)
---

I hope now you understand what I mean.

**Comment by worlpaker** [comment]
Sorry, my bad. I didn't understand the problem correctly. 
I changed the support function to this:

```json
"support_functions": {
	"comment": "Support Functions",
	"match": "(?:(?:((?<=\\.)\\w+)|(\\w+))(?=\\())",
	"captures": {
		"1": {
			"name": "entity.name.function.support.go"
		},
		"2": {
			"patterns": [
				{
					"include": "#type-declarations"
				},
				{
					"match": "\\d\\w*",
					"name": "invalid.illegal.identifier.go"
				},
				{
					"match": "\\w+",
					"name": "entity.name.function.support.go"
				}
			]
		}
	}
}
```
I got this idea from your suggested solution. I believe storage types must be simple (just keywords) to follow standards. Instead, I made that change in support functions. If there is a dot before support functions then it will not include type-declarations. Otherwise, it will include.

**v0.2.7 - Ayu Mirage Theme**
![fixed_#1](https://user-images.githubusercontent.com/34960716/229749800-186cb819-4934-43de-a0fe-b447de0a9562.png)

---

This has been fixed in v0.2.7 without occurring any other issues. But I'm open-minded about different solutions and updates.

**Comment by Vacym** [comment]
I like this solution! And yes, I agree that the storage types should be as simple as possible.
Now everything is highlighted exactly the way I wanted, and I haven't found any more unexpected problems
Thanks for being so responsive throughout this process. It was a pleasure working with you.
